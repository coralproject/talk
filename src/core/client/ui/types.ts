import React from "react";

// TODO: Extract useful common types into its own package.

/**
 * Overwrite properties of `T`.
 *
 * E.g. Omit<{a: boolean, b: boolean}, "b"> = {a: boolean}
 */
export type Omit<U, K extends keyof U> = Pick<U, Exclude<keyof U, K>>;

/**
 * Overwrite properties of `T`.
 *
 * E.g. Overwrite<{a: boolean}, {a: string}> = {a: string}
 */
export type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> & U;

/**
 * Returns the PropTypes from a React Component.
 *
 * E.g. type ButtonProps = ReturnPropTypes<Button>;
 */
export type PropTypesOf<T> = T extends React.ComponentType<infer R>
  ? R
  : T extends React.Component<infer S> ? S : {};
