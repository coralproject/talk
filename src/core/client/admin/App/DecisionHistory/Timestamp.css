$timestampColor: var(--v2-colors-mono-100);

.root {
  color: $timestampColor;
  font-family: var(--v2-font-family-primary);
  font-weight: var(--v2-font-weight-primary-bold);
  font-size: var(--v2-font-size-1);
  line-height: var(--v2-line-height-reset);
}
