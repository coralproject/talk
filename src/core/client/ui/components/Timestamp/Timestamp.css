.root {
  cursor: default;
}

.text {
  composes: timestamp from "coral-ui/shared/typography.css";
}
