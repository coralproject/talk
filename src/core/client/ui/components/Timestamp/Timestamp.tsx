import cn from "classnames";
import React, {
  EventHandler,
  FunctionComponent,
  MouseEvent,
  useCallback,
  useState,
} from "react";

import {
  AbsoluteTime,
  BaseButton,
  RelativeTime,
  UIContext,
} from "coral-ui/components";

import styles from "./Timestamp.css";

export interface TimestampProps {
  className?: string;
  children: string;
  toggleAbsolute?: boolean;
  onToggleAbsolute?: (absolute: boolean) => void;
  onClick?: EventHandler<MouseEvent>;
}

const Timestamp: FunctionComponent<TimestampProps> = props => {
  const [showAbsolute, setShowAbsolute] = useState(false);
  const handleOnClick = useCallback(
    (event: MouseEvent) => {
      if (props.toggleAbsolute) {
        if (props.onToggleAbsolute) {
          props.onToggleAbsolute(!showAbsolute);
        }
        setShowAbsolute(!showAbsolute);
      }
      if (props.onClick) {
        return props.onClick(event);
      }
    },
    [showAbsolute, setShowAbsolute, props.onClick]
  );
  return (
    <UIContext.Consumer>
      {({ locales }) => (
        <BaseButton className={styles.root} onClick={handleOnClick}>
          {showAbsolute && locales ? (
            <AbsoluteTime
              date={props.children}
              locales={locales}
              className={cn(styles.text, props.className)}
            />
          ) : (
            <RelativeTime
              className={cn(styles.text, props.className)}
              date={props.children}
            />
          )}
        </BaseButton>
      )}
    </UIContext.Consumer>
  );
};

Timestamp.defaultProps = {
  toggleAbsolute: true,
};

export default Timestamp;
