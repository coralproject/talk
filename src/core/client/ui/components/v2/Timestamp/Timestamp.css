$timestampColor: var(--v2-colors-mono-100);

.root {
  cursor: default;
}

.text {
  color: $timestampColor;
  font-family: var(--v2-font-family-primary);
  font-weight: var(--v2-font-weight-primary-regular);
  font-size: var(--v2-font-size-2);
  line-height: var(--v2-line-height-reset);
}
