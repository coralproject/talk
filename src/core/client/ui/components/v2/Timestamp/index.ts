export { default } from "./Timestamp";
export { default as TimestampFormatter } from "./TimestampFormatter";
