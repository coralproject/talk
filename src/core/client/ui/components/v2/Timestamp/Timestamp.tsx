import cn from "classnames";
import React, { FunctionComponent, useCallback, useState } from "react";

import { AbsoluteTime, BaseButton, RelativeTime } from "coral-ui/components/v2";

import styles from "./Timestamp.css";

export interface TimestampProps {
  className?: string;
  children: string;
  toggleAbsolute?: boolean;
  onToggleAbsolute?: (showAbsolute: boolean) => void;
}

const Timestamp: FunctionComponent<TimestampProps> = (props) => {
  const [showAbsolute, setShowAbsolute] = useState(false);
  const toggleShowAbsolute = useCallback(() => {
    if (props.toggleAbsolute) {
      if (props.onToggleAbsolute) {
        props.onToggleAbsolute(!showAbsolute);
      }

      setShowAbsolute(!showAbsolute);
    }
  }, [showAbsolute, setShowAbsolute, props]);
  return (
    <BaseButton className={styles.root} onClick={toggleShowAbsolute}>
      {showAbsolute ? (
        <AbsoluteTime
          date={props.children}
          className={cn(styles.text, props.className)}
        />
      ) : (
        <RelativeTime
          className={cn(styles.text, props.className)}
          date={props.children}
        />
      )}
    </BaseButton>
  );
};

Timestamp.defaultProps = {
  toggleAbsolute: true,
};

export default Timestamp;
