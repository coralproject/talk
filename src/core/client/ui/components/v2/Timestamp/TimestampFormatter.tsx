import cn from "classnames";
import React, { FunctionComponent, useCallback, useState } from "react";

import { AbsoluteTime, BaseButton, RelativeTime } from "coral-ui/components/v2";

export interface TimestampFormatterProps {
  className?: string;
  rootClassName?: string;
  textClassName?: string;
  children: string;
  toggleAbsolute?: boolean;
  onToggleAbsolute?: (showAbsolute: boolean) => void;
}

const TimestampFormatter: FunctionComponent<TimestampFormatterProps> = ({
  toggleAbsolute,
  onToggleAbsolute,
  children,
  textClassName,
  rootClassName,
  className,
}) => {
  const [showAbsolute, setShowAbsolute] = useState(false);
  const toggleShowAbsolute = useCallback(() => {
    if (toggleAbsolute) {
      if (onToggleAbsolute) {
        onToggleAbsolute(!showAbsolute);
      }

      setShowAbsolute(!showAbsolute);
    }
  }, [toggleAbsolute, onToggleAbsolute, showAbsolute, setShowAbsolute]);
  return (
    <BaseButton className={rootClassName} onClick={toggleShowAbsolute}>
      {showAbsolute ? (
        <AbsoluteTime
          date={children}
          className={cn(textClassName, className)}
        />
      ) : (
        <RelativeTime
          className={cn(textClassName, className)}
          date={children}
        />
      )}
    </BaseButton>
  );
};

TimestampFormatter.defaultProps = {
  toggleAbsolute: true,
};

export default TimestampFormatter;
