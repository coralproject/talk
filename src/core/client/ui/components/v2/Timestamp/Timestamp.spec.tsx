import React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { PropTypesOf } from "coral-framework/types";

import Timestamp from "./Timestamp";

it("renders correctly", () => {
  const props: PropTypesOf<typeof Timestamp> = {
    children: "1995-12-17T03:24:00.000Z",
  };
  const renderer = createRenderer();
  renderer.render(<Timestamp {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
