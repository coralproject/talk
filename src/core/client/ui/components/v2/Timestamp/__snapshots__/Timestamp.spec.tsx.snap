// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ForwardRef(forwardRef)
  className="Timestamp-root"
  onClick={[Function]}
>
  <ForwardRef(forwardRef)
    className="Timestamp-text"
    date="1995-12-17T03:24:00.000Z"
  />
</ForwardRef(forwardRef)>
`;
